src/components/FullscreenModal/hooks/useStickyObserver.ts

```diff
@@ -0,0 +1,59 @@
+import { useEffect, useRef } from 'react';
+import debounce from 'lodash.debounce';
+import { isNotNull } from 'ramda-adjunct';
+
+export const useStickyObserver = (
+  modalRef: React.MutableRefObject<HTMLElement>,
+  elementRef: React.MutableRefObject<HTMLElement>,
+  isInView: () => void,
+): void => {
+  const observerRef = useRef(null);
+  const listeners = useRef(null);
+  const areListenersAdded = useRef(false);
+
+  useEffect(() => {
+    const currentModalRef = modalRef.current;
+    listeners.current = {
+      resizeListener: debounce(isInView, 200),
+      scrollListener: debounce(isInView, 100),
+    };
+
+    if (areListenersAdded.current && isNotNull(listeners.current)) {
+      currentModalRef.removeEventListener(
+        'scroll',
+        listeners.current.scrollListener,
+      );
+      window.removeEventListener('resize', listeners.current.resizeListener);
+    }
+
+    if (isNotNull(observerRef.current)) {
+      observerRef.current.disconnect();
+    }
+    observerRef.current = new MutationObserver(() => isInView());
+
+    if (isNotNull(currentModalRef) && isNotNull(elementRef.current)) {
+      observerRef.current.observe(currentModalRef, {
+        childList: true,
+        subtree: true,
+        attributes: true,
+      });
+      isInView();
+
+      currentModalRef.addEventListener(
+        'scroll',
+        listeners.current.scrollListener,
+      );
+      window.addEventListener('resize', listeners.current.resizeListener);
+      areListenersAdded.current = true;
+    }
+    return () => {
+      observerRef.current.disconnect();
+      currentModalRef.removeEventListener(
+        'scroll',
+        listeners.current.scrollListener,
+      );
+      window.removeEventListener('resize', listeners.current.resizeListener);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isInView]);
```

Comment:
Does it work properly when `modalRef.current` is empty on first render? Seems like it wouldn't re-run the effect but it probably does thanks to some other cond?

Reply:
Yep, it should be working, I advise going through previous @abykovssc comments. As I understand a ref is a mutable object which exists through the whole component lifecycle and thus it doesn't need to be specified as a dependency here.

Reply:
Yeah, that's true but my point is that ref.current is empty on first render (assuming it's ref attached to a JSX comp, not just value). But probably isInView is triggered after the first render so it works 🤔

Reply:
Ok, I see https://github.com/facebook/react/issues/14387#issuecomment-503616820 . In the past I think refs attached to JSX elements were sometimes null for the first render but maybe that's not an issue anymore